File: C++/LongestSubstringNoRepeats/LongestSubstring2.cpp
/*
LeetCode 3. Longest Substring Without Repeating Characters

Given a string s, find the length of the longest substring without duplicate characters.

Example 1
Input: s = "abcabcbb"
Output: 3

Example 2
Input: s = "bbbbb"
Output: 1
*/

#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

// sliding window
// use l and r pointers, both start at beginning of string
// have an unordered_set that tracks which chars are in window
// each time we move r pointer, check if longest has grown
int longestSubstring(string s) {

    int l = 0;
    int longest = 0;
    int len = s.size();
    unordered_set<char> chars;

    for (int r = 0; r < len; r++) {

        // if we have seen char before, bump up l pointer
        char c = s[r];
        while (chars.find(c) != chars.end()) {
            chars.erase(s[l]);
            l++;
        }

        chars.insert(c);

        // check if length has grown
        longest = max(longest, r - l + 1);
    }

    return longest;
}

int main() {

    cout << "hello world" << endl;
    int result = longestSubstring("bbbbb");
    cout << "result: " << result << endl;

}

File: Python/longest_substring_2.py
def longest_substring(s: str) -> int:

    left = 0
    length = len(s)
    chars = set()
    longest = 0

    for right in range(length):
        char = s[right]

        while char in chars:
            chars.remove(s[left])
            left += 1

        chars.add(char)
        longest = max(longest, right - left + 1)

    return longest

if __name__ == "__main__":
    input = "abcabcdea"
    output = longest_substring(input)
    print(f"result: {output}")